subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/GetInputPropertiesVisitor.java

```diff
@@ -65,21 +63,18 @@ public void visitInputProperty(String propertyName, PropertyValue value, boolean
 
     @Nullable
     private static Object prepareValue(@Nullable Object value) {
-        while (true) {
-            if (value instanceof Callable) {
-                Callable callable = (Callable) value;
-                value = uncheckedCall(callable);
-            } else if (value instanceof FileCollection) {
-                FileCollection fileCollection = (FileCollection) value;
-                return fileCollection.getFiles();
-            } else {
-                return avoidGString(value);
-            }
-        }
+        Object unpacked = DeferredUtil.unpack(value);
```

Comment:
⭕️ Just a heads up that this also changes the behaviour of `prepareValue` with regards to `Provider<T>` instances. `unpack` will also `get()` values from a given a `Provider`.

Reply:
We do the provider unpacking already one level down: https://github.com/gradle/gradle/blob/8fe627eddab7012bdbcafad19b3940b4cf881407/subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/bean/AbstractNestedRuntimeBeanNode.java#L117-L124

So the new behavior would only kick in if the provider is wrapped into some lazy construct. And that is pretty wrong.